### DISCUSSION TITLE
Xunit.v3 transitive dependency and test discovery

### DISCUSSION THREAD

**Discussion by dnv-kimbell** [xunit/xunit#3074] [General]
I maintain shared libraries and testing infrastructure used by a large number of internal applications. For historical reasons, multiple testing frameworks have been used (per application)

One of the things we have done is creating our internal testing packages for XUnit, NUnit, etc that controls the dependencies on the relevant external packages. Our applications take a dependency on these and get things through transitive dependencies.

In XUnit 2 this worked without any problem. 
I've started experimenting with XUnit 3 and tests did not get discovered. During trial and error, I added the Xunit.v3 package as a direct dependency on my testing project and then tests are discovered and run.

Is this change by design or by accident? It would be nice to keep the way we control versions, but it's not a big deal in the great scheme of things.

**Comment by bradwilson** [comment]
This change is not by design. Can you provide a repro?

Generally speaking, when people write extensions to xUnit.net v3 they reference `xunit.v3.extensibility.core` and not `xunit.v3`, since the latter will bring in all the machinery necessary to auto-generate an executable (which you don't want when creating an extension library). Maybe this is the answer to your problem?

**Comment by dnv-kimbell** [comment]
With V2 it was a convenient way of ensuring applications used a relatively new version.  I created a repro https://github.com/dnv-kimbell/xunit-transitive

As I was setting things up, I used dotnet test instead of Visual Studio. When I use a transitive dependency, I get an exception pointing to the thing you describe; auto-generated executable. This makes sense based on the new testing platform, and we will just have to rethink how we are setting this up. 

Thanks for the quick answer. 

This is the exception we get
```
Restore complete (0,9s)
  Dependency succeeded (0,1s) → Dependency\bin\Debug\net9.0\Dependency.dll
  TransitiveRepro succeeded with 1 warning(s) (0,9s) → TransitiveRepro\bin\Debug\net9.0\TransitiveRepro.dll
    C:\Github\xunit-transitive\TransitiveRepro\TransitiveRepro\obj\Debug\net9.0\TestPlatformEntryPoint.cs(10,66): warning CS8892: Method 'TestingPlatformEntryPoint.Main(string[])' will not be used as an entry point because a synchronous entry point 'AutoGeneratedProgram.Main(string[])' was found.
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v3.0.0-pre.49+a7f5dc0cba (64-bit .NET 9.0.0)
[xUnit.net 00:00:00.45] TransitiveRepro: Catastrophic failure: System.InvalidOperationException: Test process terminated unexpectedly.
   at Xunit.Runner.v3.Xunit3..ctor(XunitProjectAssembly projectAssembly, ISourceInformationProvider sourceInformationProvider, IMessageSink diagnosticMessageSink, ITestProcessLauncher testProcessLauncher)
   at Xunit.Runner.v3.Xunit3.ForDiscoveryAndExecution(XunitProjectAssembly projectAssembly, ISourceInformationProvider sourceInformationProvider, IMessageSink diagnosticMessageSink, ITestProcessLauncher testProcessLauncher)
   at Xunit.XunitFrontController.Create(XunitProjectAssembly projectAssembly, ISourceInformationProvider sourceInformationProvider, IMessageSink diagnosticMessageSink, ITestProcessLauncher testProcessLauncher)
   at Xunit.Runner.VisualStudio.VsTestRunner.RunTestsInAssembly(IRunContext runContext, IFrameworkHandle frameworkHandle, LoggerHelper logger, TestPlatformContext testPlatformContext, RunSettings runSettings, IMessageSink reporterMessageHandler, AssemblyRunInfo runInfo)
Additionally, path to test adapters can be specified using /TestAdapterPath command. Example  /TestAdapterPath:<pathToCustomAdapters>.
  TransitiveRepro test failed with 1 error(s) and 1 warning(s) (1,7s)
    C:\Program Files\dotnet\sdk\9.0.100\Microsoft.TestPlatform.targets(48,5): warning : No test is available in C:\Github\xunit-transitive\TransitiveRepro\TransitiveRepro\bin\Debug\net9.0\TransitiveRepro.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
    C:\Program Files\dotnet\sdk\9.0.100\Microsoft.TestPlatform.targets(48,5): error MSB6006: "dotnet.exe" exited with code 1.

Test summary: total: 0; failed: 0; succeeded: 0; skipped: 0; duration: 1,7s
Build failed with 1 error(s) and 2 warning(s) in 4,3s
```
